File: Camp Progress/695. Max Area of Island.cpp
// Question Link: https://leetcode.com/problems/max-area-of-island

class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int M = grid.size();
        int N = grid[0].size();
        int maxAreaIsland = 0;
        
        unordered_map<string, string> dsu;
        vector<vector<int>> ranks(M, vector<int>(N, 1));
        vector<int> dirs = { 0, 1, 0, -1, 0 };
        
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == 1) {
                    maxAreaIsland = max(maxAreaIsland, ranks[i][j]);
                    for (int d = 1; d < 5; d++) {
                        int ni = dirs[d] + i;
                        int nj = dirs[d - 1] + j;
                        if (isInside(ni, nj, M, N) && grid[ni][nj] == 1) {
                            Union(i, j, ni, nj, dsu, ranks, maxAreaIsland);
                        }
                    }
                }
            }   
        }
        
        return maxAreaIsland;
    }
    
    void Union(int ai, int aj, int bi, int bj, unordered_map<string, string> &dsu, vector<vector<int>> &ranks, int &maxAreaIsland) {
        
        string a = getKey(ai, aj);
        string b = getKey(bi, bj);
        
        dsu[a] = (dsu[a] == "") ? a : dsu[a];
        dsu[b] = (dsu[b] == "") ? b : dsu[b];
        
        string parentA = find(a, dsu);
        string parentB = find(b, dsu);
        
        if (parentA == parentB) {
            return;
        }
        
        dsu[parentB] = parentA;
        
        vector<int> parA = getParentCoordinates(parentA);
        vector<int> parB = getParentCoordinates(parentB);
        
        ranks[parA[0]][parA[1]] += ranks[parB[0]][parB[1]];   
        
        maxAreaIsland = max({ maxAreaIsland, ranks[parA[0]][parA[1]], ranks[parB[0]][parB[1]] });
    }
    
    string find(string cur, unordered_map<string, string> &dsu) {
        if (dsu[cur] == cur) {
            return cur;
        }
        return dsu[cur] = find(dsu[cur], dsu);
    }
    
    bool isInside(int i, int j, int &M, int &N) {
        if (i < 0 || i >= M || j < 0 || j >= N) {
            return false;
        }
        return true;
    }
    
    vector<int> getParentCoordinates(string &parent) {
        int pos = 0;
        for (; pos < parent.size(); pos++) {
            if (parent[pos] == ',') {
                break;
            }
        }
        string i = parent.substr(0, pos), j = parent.substr(pos + 1);
        return { stoi(i), stoi(j) };
    }
    
    string getKey(int i, int j) {
        return to_string(i) + "," + to_string(j);
    }
};

File: Python3 Solutions/695. Max Area of Island.py
# Question Link: https://leetcode.com/problems/max-area-of-island/

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        maxIslands = [0]
        
        dirs = [0, 1, 0, -1, 0]
        dsu = {}
        rank = [[1]*n for _ in range(m)]
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    maxIslands[0] = max(maxIslands[0], grid[i][j])
                    for d in range(1, 5):
                        ni, nj = dirs[d] + i, dirs[d - 1] + j
                        if self.is_valid(ni, nj, m, n, grid):
                            a = str(i)+'.'+str(j)
                            b = str(ni)+'.'+str(nj)
                            self.union(a, b, dsu, rank, maxIslands)
                        
        return maxIslands[0]
        
    def union(self, a, b, dsu, rank, maxIslands):
        dsu[a] = dsu[a] if a in dsu else a
        dsu[b] = dsu[b] if b in dsu else b
        
        A = self.find(a, dsu)
        B = self.find(b, dsu)
        
        if A == B:
            return
        
        dsu[B] = A
        
        Ai, Aj = list(map(int, A.split('.')))
        Bi, Bj = list(map(int, B.split('.')))
        
        rank[Ai][Aj] = rank[Ai][Aj] + rank[Bi][Bj]
        
        maxIslands[0] = max(maxIslands[0], rank[Ai][Aj])
        
    def find(self, c, dsu):
        if dsu[c] == c:
            return c
        dsu[c] = self.find(dsu[c], dsu)
        return dsu[c]
        
    def is_valid(self, i, j, m, n, grid):
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == 0:
            return False
        return True
        
        